### PR TITLE
Adaptive track selection

### PR DESCRIPTION
**Pull request by Mecharyry**
## Problem
Clients have been experiencing a long buffer period and occasionally `BehindLiveWindowException` which occurs when the video playback cannot keep pace with the live stream.

These problems are all associated with `no-player` playing the highest quality video representation all of the time 😱  We should be adaptively switching between representations based on the bandwidth available to play content.

## Solution
Make use of the `DefaultBandwidthMeter` passing it into both the `DefaultDataSourceFactory` and the `DefaultTrackSelector`.

Adding it to the `DefaultDataSourceFactory` attaches it to any media that is fetched in order to acquire the estimated bitrate. Passing to the `DefaultTrackSelector` ensures that tracks are switched based on this bitrate.

The following logs show the switching of tracks based on bitrate.

```
09-14 15:23:07.644 20141-20199/com.novoda.demo E/FOO: bitrateEstimate: 1854275
09-14 15:23:07.644 20141-20199/com.novoda.demo E/FOO: effectiveBitrate: 1390706
09-14 15:23:07.644 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:07.644 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:07.644 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:07.644 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate true
09-14 15:23:07.644 20141-20199/com.novoda.demo E/FOO: selectedIndex: 3
09-14 15:23:26.169 20141-20199/com.novoda.demo E/FOO: bitrateEstimate: 2234009
09-14 15:23:26.169 20141-20199/com.novoda.demo E/FOO: effectiveBitrate: 1675506
09-14 15:23:26.169 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:26.169 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:26.169 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:26.169 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate true
09-14 15:23:26.169 20141-20199/com.novoda.demo E/FOO: selectedIndex: 3
09-14 15:23:28.837 20141-20199/com.novoda.demo E/FOO: bitrateEstimate: 4298112
09-14 15:23:28.837 20141-20199/com.novoda.demo E/FOO: effectiveBitrate: 3223584
09-14 15:23:28.837 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:28.837 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:28.837 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate true
09-14 15:23:28.837 20141-20199/com.novoda.demo E/FOO: selectedIndex: 2
09-14 15:23:50.172 20141-20199/com.novoda.demo E/FOO: bitrateEstimate: 8757550
09-14 15:23:50.172 20141-20199/com.novoda.demo E/FOO: effectiveBitrate: 6568162
09-14 15:23:50.172 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:50.172 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:50.172 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate true
09-14 15:23:50.172 20141-20199/com.novoda.demo E/FOO: selectedIndex: 2
09-14 15:23:53.094 20141-20199/com.novoda.demo E/FOO: bitrateEstimate: 8884962
09-14 15:23:53.094 20141-20199/com.novoda.demo E/FOO: effectiveBitrate: 6663721
09-14 15:23:53.094 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:53.094 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:23:53.094 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate true
09-14 15:23:53.094 20141-20199/com.novoda.demo E/FOO: selectedIndex: 2
09-14 15:24:14.166 20141-20199/com.novoda.demo E/FOO: bitrateEstimate: 17150580
09-14 15:24:14.166 20141-20199/com.novoda.demo E/FOO: effectiveBitrate: 12862935
09-14 15:24:14.167 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate false
09-14 15:24:14.167 20141-20199/com.novoda.demo E/FOO: format.bitrate <= effectiveBitrate true
09-14 15:24:14.167 20141-20199/com.novoda.demo E/FOO: selectedIndex: 1
```

### Test(s) added 
No, just updating a small piece of creation logic. We have tested this both in `no-player` and a client application.

### Screenshots
No UI changes.

### Paired with 
@zegnus

### CODE REVIEW

File: core/src/main/java/com/novoda/noplayer/internal/exoplayer/NoPlayerExoPlayerCreator.java [hunk 1]
```diff
@@ -47,10 +50,12 @@ public NoPlayer createExoPlayer(Context context, DrmSessionCreator drmSessionCre
         }
 
         ExoPlayerTwoImpl create(Context context, DrmSessionCreator drmSessionCreator, boolean downgradeSecureDecoder) {
-            DefaultDataSourceFactory defaultDataSourceFactory = new DefaultDataSourceFactory(context, "user-agent");
+            DefaultBandwidthMeter bandwidthMeter = new DefaultBandwidthMeter();
```
Comment (zegnus) [inline core/src/main/java/com/novoda/noplayer/internal/exoplayer/NoPlayerExoPlayerCreator.java]: ❤️

**Comment by juankysoriano** [comment]
💯 very nice!

**Comment by juankysoriano** [comment]
Is this a feature exclusive of `3.0.1` ? or also available in the previous version? If we are upgrading to `3.0.1` could we show the changelog?

**Comment by Mecharyry** [comment]
> If we are upgrading to 3.0.1 could we post the changelog?
This is our version number, I'm upping it so that we can perform a release.

This feature has always been in the latest version of ExoPlayer, we just forgot to add it in!

**Pull request merged**